### PR TITLE
bpo-44589: raise SyntaxError when mapping pattern has duplicate key

### PR DESCRIPTION
**Pull request by jdevries3133**
* match case pattern cannot have duplicate keys per PEP 634.
* when duplicate literal values are present, a syntax error should be
  raised; not a runtime ValueError, which was the prior behavior.

Tests have been updated to test the new behavior. Docs were not changed, because the behavior is now more closely aligned with the documentation and the PEP.

https://bugs.python.org/issue44589

### CODE REVIEW

**Comment by jdevries3133** [comment]
I switched this to a draft because I realized that I still need to test the edge cases that Brandt pointed out on the bpo:

> We'll want to make sure we have test coverage of edge cases like {0: _, False: _}, {0: _, 0.0: _}, {0: _, -0: _}, {0: _, 0j: _}, etc.

I will add those when I get the chance! In the mean time, I'd appreciate feedback on what I have so far if anyone stops by.

File: Python/compile.c [hunk 1]
```diff
@@ -6183,12 +6190,26 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
             SET_LOC(c, ((pattern_ty) asdl_seq_GET(patterns, i)));
             return compiler_error(c, e);
         }
+
+        if (key->kind == Constant_kind) {
+            if (PySet_Contains(seen, key->v.Constant.value)) {
+                const char *e =  "Duplicate literal keys in match pattern are "
+                                 "not allowed";
+                return compiler_error(c, e);
```
Comment (Fidget-Spinner) [inline Python/compile.c]: I think you need to call ``Py_DECREF(seen);`` too. The are also other 2 other ``compiler_error`` paths and you may need to decref too. Otherwise errors will leak memory.

If you want a cleaner implementation, maybe ``goto error`` and decref and ``return compiler_error(c, e);`` at that label. That way each branch only needs to set ``e`` then ``goto error``.

**Review by Fidget-Spinner — COMMENTED** [review]
Looks good to me in general. I'm not a compiler expert though. Thanks for taking this up :).

**Comment by Fidget-Spinner** [comment]
The docs at https://docs.python.org/3.11/reference/compound_stmts.html#mapping-patterns also need fixing.
Specifically the lines:
> Duplicate key values in mapping patterns are disallowed. (If all key patterns are literal patterns this is considered a syntax error; otherwise this is a runtime error and will raise ValueError.)

And

> If duplicate keys are detected in the mapping pattern, the pattern is considered invalid and ValueError is raised.

require more clarification.

**Comment by jdevries3133** [comment]
@Fidget-Spinner re: Documentation, I tweaked the docs. See aeb4c42b4f67d688069a6e981335e661e4b3006c

File: Python/compile.c (hunk 1 above)
Comment (jdevries3133) [reply Python/compile.c]: Thank you, I really appreciate that feedback. See 3a96d32b94291ac9dcf5ba5cc9c09ed5a5f692f2 for the changes I made in response. I decided to not to use the `goto` pattern because different exit branches are doing different things. I could probably make it work with a `goto compiler_error` and `goto resource_error`, since those seem to be the two ways that this function exits, but then there is also the variadic call to `compiler_error` to think about if we use `goto`'s:

https://github.com/python/cpython/blob/082ded0e7672777e5ee0168cd1759a53b8c3f813/Python/compile.c#L6151-L6152

What do you think? Is it best to not use `goto`, all things considered?

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Great job!

File: Python/compile.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Python/compile.c]: Sorry I wasn't being very clear. I meant to consider using ``goto`` just for those 3 paths where you need to decref `seen`. Usually if there needs to be cleanup (like a decref), the convention is to goto ``error`` so that the cleanup code is grouped together and not all over the place. You can probably find this pattern in much of the C code in CPython.

However, since this code isn't too complex (yet). It's probably not needed. I'll just wait and see what the core devs have to say. Personally, I am fine with the code as-is.

File: Python/compile.c [hunk 2]
```diff
@@ -6175,20 +6175,44 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
     }
     // Collect all of the keys into a tuple for MATCH_KEYS and
     // COPY_DICT_WITHOUT_KEYS. They can either be dotted names or literals:
+
+    // will hold only Constant_kind keys
+    PyObject *seen = PySet_New(NULL);
+    if (seen == NULL) {
+        return -1;
```
Comment (brandtbucher) [inline Python/compile.c]: All of the visitors in the compiler return `0` on failure and `1` on success:

```suggestion
        return 0;
```

File: Python/compile.c [hunk 3]
```diff
@@ -6175,20 +6175,44 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
...
         }
+
+        if (key->kind == Constant_kind) {
+            if (PySet_Contains(seen, key->v.Constant.value)) {
+                const char *e =  "Duplicate literal keys in match pattern are "
+                                 "not allowed";
+                Py_DECREF(seen);
+                return compiler_error(c, e);
+            }
+            PySet_Add(seen, key->v.Constant.value);
```
Comment (brandtbucher) [inline Python/compile.c]: This needs to be checked for errors:

```suggestion
            if (PySet_Add(seen, key->v.Constant.value)) {
                Py_DECREF(seen);
                return 0;
            }
```

File: Python/compile.c [hunk 4]
```diff
@@ -6175,20 +6175,44 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
...
         if (key == NULL) {
             const char *e = "can't use NULL keys in MatchMapping "
                             "(set 'rest' parameter instead)";
             SET_LOC(c, ((pattern_ty) asdl_seq_GET(patterns, i)));
+            Py_DECREF(seen);
             return compiler_error(c, e);
         }
+
+        if (key->kind == Constant_kind) {
+            if (PySet_Contains(seen, key->v.Constant.value)) {
```
Comment (brandtbucher) [inline Python/compile.c]: This needs to be checked for errors:

```suggestion
            int in_seen = PySet_Contains(seen, key->v.Constant.value);
            if (in_seen < 0) {
                Py_DECREF(seen);
                return 0;
            }
            if (in_seen) {
```

File: Python/compile.c [hunk 5]
```diff
@@ -6175,20 +6175,44 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
...
+                const char *e =  "Duplicate literal keys in match pattern are "
+                                 "not allowed";
```
Comment (brandtbucher) [inline Python/compile.c]: I'd like to make this error message the same as the runtime error message. We can also provide a hint to the user by clarifying *which* key is duplicated:

```suggestion
                const char *e =  "mapping pattern checks duplicate key (%R)";
```

You can then format the message by adding `key->v.Constant.value` to the `compiler_error` call below.

File: Python/compile.c (hunk 1 above)
Comment (brandtbucher) [reply Python/compile.c]: Please use `goto error;`. With my newest suggestions, I think that the amount of repeated error-handling code is becoming burdensome.

Your error block at the end of this function can just be:
```c
error:
    Py_DECREF(seen);
    return 0;
```

And then all of these error branches just become something like:

```c
const char *e = "...";
compiler_error(c, e);
goto error;
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]

File: Python/compile.c (hunk 3 above)
Comment (Fidget-Spinner) [reply Python/compile.c]: Wow thanks for pointing the set error check requirements out. I'd forgotten about them and learnt something new today :).

**Comment by jdevries3133** [comment]
Thank you very much for your feedback!! I have made the requested changes; please review again.

File: Python/compile.c [hunk 6]
```diff
@@ -6173,22 +6173,52 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
     if (INT_MAX < size - 1) {
         return compiler_error(c, "too many sub-patterns in mapping pattern");
     }
+
```
Comment (brandtbucher) [inline Python/compile.c]: Minor: please revert this unrelated whitespace.

```suggestion
```

File: Python/compile.c [hunk 7]
```diff
@@ -6173,22 +6173,52 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
...
+                const char *e = "mapping pattern checks duplicate key (%R)";
+                compiler_error(c, e, key->v.Constant.value);
+                goto error;
+            }
+            if (PySet_Add(seen, key->v.Constant.value)) {
+                goto error;
+            }
+        }
+
         if (!MATCH_VALUE_EXPR(key)) {
```
Comment (brandtbucher) [inline Python/compile.c]: We don't really need that macro here anymore:

```suggestion
        else if (key->kind != Attribute_kind) {
```

File: Python/compile.c [hunk 8]
```diff
@@ -6173,22 +6173,52 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
...
+
     // Collect all of the keys into a tuple for MATCH_KEYS and
     // COPY_DICT_WITHOUT_KEYS. They can either be dotted names or literals:
+
+    // will hold only Constant_kind keys
+    PyObject *seen = PySet_New(NULL);
+    if (seen == NULL) {
+        return 0;
+    }
+
```
Comment (brandtbucher) [inline Python/compile.c:6185]: Maybe add a comment to the effect of:

```suggestion
    // NOTE: In this loop, goto error on failure. Otherwise we'll leak seen!
```

File: Misc/NEWS.d/next/Core and Builtins/2021-07-13-23-19-41.bpo-44589.59OH8T.rst [hunk 9]
```diff
@@ -0,0 +1,2 @@
+Mapping patterns in match/case statements with two or more identical literal
+keys causes a syntax error.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2021-07-13-23-19-41.bpo-44589.59OH8T.rst]:
```suggestion
Mapping patterns in ``match`` statements with two or more equal literal
keys will now raise a :exc:`SyntaxError` at compile-time.
```

**Review by brandtbucher — APPROVED** [review]
Nice, thanks for the quick changes.

I have a few minor suggestions, but I consider this merge-ready and will land it EOD regardless.

**Comment by jdevries3133** [comment]
@brandtbucher ok, those final changes have all been made in 00886a57e05696427f813bbe4c27854b9e347409

File: Python/compile.c [hunk 10]
```diff
@@ -6175,20 +6175,51 @@ compiler_pattern_mapping(struct compiler *c, pattern_ty p, pattern_context *pc)
...
         }
-        if (!MATCH_VALUE_EXPR(key)) {
+
+        else if (key->kind != Attribute_kind) {
             const char *e = "mapping pattern keys may only match literals and attribute lookups";
-            return compiler_error(c, e);
+            compiler_error(c, e);
+            goto error;
         }
         VISIT(c, expr, key);
```
Comment (brandtbucher) [inline Python/compile.c]: Ah, wait, this macro was hiding one more leak:

```suggestion
        if (!compiler_visit_expr(c, key)) {
            goto error;
        }
```

**Review by brandtbucher — CHANGES_REQUESTED** [review]

File: Python/compile.c (hunk 10 above)
Comment (jdevries3133) [reply Python/compile.c]: Wow, awesome catch; that's a sneaky one. Resolved in a8e6751cd663ee27548221345b20b4295367b984

**Comment by jdevries3133** [comment]
I have made the requested changes; please review again.
